sensor/tests/resource/service/service_test.go

```diff
@@ -0,0 +1,352 @@
+package service
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/utils"
+	"github.com/stackrox/rox/sensor/tests/resource"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"sigs.k8s.io/e2e-framework/klient/k8s"
+)
+
+var (
+	NginxDeployment          = resource.YamlTestFile{Kind: "Deployment", File: "nginx.yaml"}
+	NginxPod                 = resource.YamlTestFile{Kind: "Deployment", File: "nginx-pod.yaml"}
+	NginxServiceClusterIP    = resource.YamlTestFile{Kind: "Service", File: "nginx-service-cluster-ip.yaml"}
+	NginxServiceNodePort     = resource.YamlTestFile{Kind: "Service", File: "nginx-service-node-port.yaml"}
+	NginxServiceLoadBalancer = resource.YamlTestFile{Kind: "Service", File: "nginx-service-load-balancer.yaml"}
+)
+
+func GetLastMessageWithDeploymentName(messages []*central.MsgFromSensor, n string) *central.MsgFromSensor {
+	var lastMessage *central.MsgFromSensor
+	for i := len(messages) - 1; i >= 0; i-- {
+		if messages[i].GetEvent().GetDeployment().GetName() == n {
+			lastMessage = messages[i]
+			break
+		}
+	}
+	return lastMessage
+}
+
+func GetLastAlertsWithDeploymentID(messages []*central.MsgFromSensor, id string) *central.MsgFromSensor {
+	var lastMessage *central.MsgFromSensor
+	for i := len(messages) - 1; i >= 0; i-- {
+		if messages[i].GetEvent().GetAlertResults().GetDeploymentId() == id {
+			lastMessage = messages[i]
+			break
+		}
+	}
+	return lastMessage
+}
+
+func assertLastDeploymentHasPortExposure(t *testing.T, messages []*central.MsgFromSensor, ports []*storage.PortConfig, alerts []*storage.Alert) {
+	lastNginxDeploymentUpdate := GetLastMessageWithDeploymentName(messages, "nginx-deployment")
+	lastNginxDeploymentAlerts := GetLastAlertsWithDeploymentID(messages, lastNginxDeploymentUpdate.GetEvent().GetDeployment().GetId())
+	require.NotNil(t, lastNginxDeploymentUpdate, "should have found a message for nginx-deployment")
+	require.NotNil(t, lastNginxDeploymentAlerts, "should have found an alert for nginx-deployment")
+	deployment := lastNginxDeploymentUpdate.GetEvent().GetDeployment()
+	actualAlerts := lastNginxDeploymentAlerts.GetEvent().GetAlertResults().GetAlerts()
+	for _, expectedAlert := range alerts {
+		foundAlert := false
+		for _, actualAlert := range actualAlerts {
+			if expectedAlert.GetPolicy().GetName() == actualAlert.GetPolicy().GetName() {
+				assert.Equal(t, expectedAlert.GetState(), actualAlert.GetState())
+				foundAlert = true
+			}
+		}
+		assert.True(t, foundAlert, "Alert not found")
+	}
+
+	for _, expectedPort := range ports {
```

Comment:
I think you can rework this nested for loop to use `t.Equal` since it supports deep equal. It compares values and not memory addresses, so it would work comparing two slices with struct pointers.

Here's an example: https://go.dev/play/p/j7mZhEDP4kx

There are two caveats:
- Expected struct must be exactly the same. It wouldn't work if you're checking that only one property is there
- Order matters. But this can be easily worked around if you sort both `[]*storage.PortConfig` slices by the same property.

Reply:
The problem with using the deep equal is that `PortConfig_ExposureInfor` contains fields that vary with each test (e.g. `ServiceId`) therefore it will fail every time.